### PR TITLE
[MRG] Add support for indexing/slicing Annotations objects

### PR DESCRIPTION
**Pull request by massich**
Based on [python 3 doc](https://docs.python.org/3/reference/datamodel.html#object.__getitem__) indexing with
the wrong type should raise `TypeError` not `IndexError`. Then the entire function gets much more simpler:

``` python

def __getitem__(self, key):
    """Propagate indexing and slicing to the underlying numpy structure."""
    return (self.onset[key], self.duration[key], self.description[key])
```

About the current solution: 
* (pros) I like the idea that we are raising the right 
* (cons) the code has 5 branches

### CODE REVIEW

**Comment by massich** [comment]
based on https://github.com/mne-tools/mne-python/pull/5795#discussion_r242162577 maybe we should return a copy. But I'm not sure. 

But I guess that @larsoner was trying to avoid this?
```py
my_wrong_recorded_annotatoins = [d.startswith('foo') for d in raw.annotations.description]
onsets, _, _ = raw.annotations[my_wrong_recorded_annotatoins]
onsets += 10
```
If you want to do that you should do `raw.annotations.onset[my_wrong_recorded_annotations] += 10`

**Comment by agramfort** [comment]
+1 to return a copy

>

**Comment by larsoner** [comment]
Yes in MNE we (should, at least) always return a copy with indexing operations on our objects. This makes us different from NumPy, which has inplace and copy rules.

File: tutorials/plot_object_annotations.py [hunk 1]
```diff
@@ -189,6 +180,47 @@
...
 ###############################################################################
+# Iterating, Indexing and Slicing :class:`mne.Annotations`
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# :class:`~mne.Annotations` supports the same indexing and slicing as
+# ``numpy.array``, and it returns a ``(onset, duration, description)`` tuple
+# with the sliced elements.
+#
+# See the following examples and usages:
+plt.close('all')
```
Comment (agramfort) [inline tutorials/plot_object_annotations.py]: why matplotlib here?

File: tutorials/plot_object_annotations.py [hunk 2]
```diff
@@ -189,6 +180,47 @@
...
+print('type: {0} content: {1}'.format(type(annot[:]), annot[:]))
+
+_MSG = 'annotation #{0}: onset={1} duration={2} desc={3}'
+for idx, (onset, duration, desc) in enumerate(annot):
+    print(_MSG.format(idx, onset, duration, desc))
+
+start, stop, step = (0, None, 2)
+every_other_annotation = slice(start, stop, step)
+for onset, duration, desc in zip(*annot[every_other_annotation]):
+    print('onset={0} duration={1} desc={2}'.format(onset, duration, desc))
```
Comment (agramfort) [inline tutorials/plot_object_annotations.py]: I would try to dumb it down. Here you use advanced string formatting, for loops etc. I would just do:

```
annotations[:3]   # will return a new Annotations formed by the first 3
annotations[2]   # will return a new Annotations restricted to the 3rd annotation.
```

and I would point to python indexing doc as it behaves likes for str or lists etc.

File: tutorials/plot_object_annotations.py [hunk 3]
```diff
@@ -1,27 +1,12 @@
 """
-The **events** and :class:`~mne.Annotations` data structures
+The **:term:`<events>`** and :class:`~mne.Annotations` data structures
 =========================================================================
 
-Events and :class:`~mne.Annotations` are quite similar.
+:term:`Events <events>` and :term:`annotations` are quite similar.
```
Comment (agramfort) [inline tutorials/plot_object_annotations.py]: term links work?

File: tutorials/plot_object_annotations.py (hunk 1 above)
Comment (massich) [reply tutorials/plot_object_annotations.py]: This should go.

File: tutorials/plot_object_annotations.py (hunk 2 above)
Comment (massich) [reply tutorials/plot_object_annotations.py]: Thats exactly what does not wat I expected from the tests.

File: tutorials/plot_object_annotations.py (hunk 3 above)
Comment (massich) [reply tutorials/plot_object_annotations.py]: it should

**Comment by massich** [comment]
I confused `__getitem__` and `__iter__` behavior.

File: mne/annotations.py [hunk 4]
```diff
@@ -214,6 +214,46 @@ def __iadd__(self, other):
...
+                out = Annotations(onset=self.onset[key],
+                                  duration=self.duration[key],
+                                  description=self.description[key],
+                                  orig_time=self.orig_time)
+            elif isinstance(key, int):
+                print('__getitem__ with integer: {}'.format(key))
+                out = Annotations(onset=[self.onset[key]],
+                                  duration=[self.duration[key]],
+                                  description=[self.description[key]],
+                                  orig_time=self.orig_time)
```
Comment (agramfort) [inline mne/annotations.py]: I would return a tuple like

(self.onset[key], self.duration[key], self.description[key])

this would be consistent with the iterator.

or it could be a dict with keys onset, duration, description like when you index a pandas dataframe...

thoughts @jona-sassenhagen ?

Comment (massich) [reply mne/annotations.py]: After talking to @agramfort offline he convinced me to return a dictionary not an `Annotations` when indexing with a single integer.

It has the advantage that if we ever want to extend the annotations, with more fields we won't break people's code.

What do you think?

Comment (larsoner) [reply mne/annotations.py]: Why don't we return an Annotations object for the iterator and for slicing/indexing operations? It has the same extend-ability because whatever we add will be attributes of the Annotations class immediately.

This is also basically what we do with the Epochs class, so it's more consistent with that, too.

Comment (massich) [reply mne/annotations.py]: @mne-tools/contributors ?

Comment (jona-sassenhagen) [reply mne/annotations.py]: So the proposals are to return:

- annotations object
- tuple
- dict

?

I haven't worked with annotations enough to have an opinion ...

Comment (larsoner) [reply mne/annotations.py]: Yep, so concretely:
```
# annot
for a in raw.annotations:
    onset, duration, desc = a.onset[0], a.duration[0], a.desc[0]
    ...
# tuple
for onset, duration, desc in raw.annotations:
    ...
# dict
for a in raw.annotations:
    onset, duration, desc = a['onset'], a['duration'], a['desc']
    ...
```
The only advantage of an `dict` approach over iterating over annotations objects I see is that it avoids the `[0]`, but this does not seem worth introducing more API inconsistency to me.

Comment (larsoner) [reply mne/annotations.py]: The `annot` will also not break if you start adding a channel name for annotations.

So the question is, do we value internal package consistency (Annotations iterating like Epochs) or consistency with what Pandas does here?

Comment (larsoner) [reply mne/annotations.py]: I think the argument you are really trying to make has to do with how conceptualize iteration over an N-dimensional object: is what you get in the loop (N-1)D (like arrays, lists, tuples, pandas, etc.) or ND (like epochs). I can see why this would be useful, even if it's less like what we usually do, so I can live with it.

(FWIW, I don't quite see why the iterable argument helps, since it actually holds already for the `epochs` object (`epochs[k] == e for k, e in enumerate(epochs)`) and would hold for the `annotations` API (`annot[k] == a for k, a in annotations`). It seems to actually favor "iter yields Annotations" if `__getitem__` always returns `Annotations`; in order for the relationship to hold for "iter yields dict", we'd need `Annotations | dict` to be returned by `__getitem__`, depending on whether the result is slice vs int...)

**Comment by agramfort** [comment]
it does not hold for epochs class. if you iterate over epochs you get
arrays but when you do epochs[k] you get an epochs object

**Comment by larsoner** [comment]
Ahh true I thought it was the other way but did not check.

FWIW the iter/index equivalence argument still send to go against the proposal, though, right?

**Comment by agramfort** [comment]
> Ahh true I thought it was the other way but did not check.
>
> FWIW the iter/index equivalence argument still send to go against the proposal, though, right?

I don't think so. I propose to get a dictionary when you iter or
access k'th element.

**Comment by larsoner** [comment]
So in getitem, int gives dict and slice gives Annotations object?

**Comment by larsoner** [comment]
Or I guess we could have slice give a dict with values that are 2D ndarray. Then if you want Annotations it's just a reconstruction with double star away

**Comment by agramfort** [comment]
slice gets you Annotations and int gets you a dict

clear?

>

**Comment by larsoner** [comment]
I can live with it

**Comment by agramfort** [comment]
:)

@massich ok for you?

>

**Comment by massich** [comment]
Yes!!

I'm AFK. I'll catch up

File: mne/annotations.py [hunk 5]
```diff
@@ -214,6 +214,37 @@ def __iadd__(self, other):
                                                  other.orig_time))
         return self.append(other.onset, other.duration, other.description)
 
+    def __iter__(self):
+        for idx in range(len(self.onset)):
+            yield (self.onset[idx], self.duration[idx], self.description[idx])
```
Comment (agramfort) [inline mne/annotations.py]: I thought we said we wanted to use dict now?

File: mne/annotations.py [hunk 6]
```diff
@@ -214,6 +214,37 @@ def __iadd__(self, other):
...
+
+    def __getitem__(self, key):
+        """Propagate indexing and slicing to the underlying numpy structure."""
+        try:
+            if isinstance(key, slice) or isinstance(key, list):
```
Comment (agramfort) [inline mne/annotations.py]: I would do:

onset=self.onset[key]

and then test that if onset.ndim = 1 then return a dict. It avoids the try complication

File: mne/annotations.py (hunk 5 above)
Comment (massich) [reply mne/annotations.py]: I understand and agree with using dict when slicing with a single element. 

> slice gets you Annotations and int gets you a dict

Mainly because if we return a tuple then we break backward compatibility if we change it later.

But I missed that you were also proposing to change the iterator. The entire purpose of this PR was to be able to rewrite 

https://github.com/mne-tools/mne-python/blob/68410a1468b7fa5d72ec631a844a1b448caf8223/mne/annotations.py#L798-L802

into 

```py
for onset, duration, description in annotations[event_sel]:
    # do stuff
```

but you cannot do `for onset, duration, description in annotations[event_sel].values()`

But I might be doing something wrong

```py
    class Ob_tuple(object):
        def __init__(self):
            self.NUM = 5
            self.a = np.array([x for x in range(self.NUM)])
            self.b = np.array([x.__repr__() for x in range(self.NUM)])

        def __iter__(self):
            for ii in range(self.NUM):
                yield (self.a[ii], self.b[ii])

    for a, b in Ob_tuple():
        print((a, b))

    class Ob_dict(object):
        def __init__(self):
            self.NUM = 5
            self.a = np.array([x for x in range(self.NUM)])
            self.b = np.array([x.__repr__() for x in range(self.NUM)])

        def __iter__(self):
            for ii in range(self.NUM):
                yield {'a': self.a[ii], 'b': self.b[ii]}

    # the dict elements are ok
    for x in Ob_dict():
        print(x)

    # iterate over a dict gives you the keys not the values
    for a, b in Ob_dict():
        print((a, b))
```

**Comment by agramfort** [comment]
For any standard python object accessing to element k is like iterating k times. A mix of tuple and dict would be weird

File: mne/annotations.py (hunk 6 above)
Comment (massich) [reply mne/annotations.py]: The complication of the trys is just to raise `TypeError` instead of `IndexError` when due. (Following the standard recommendations)

and the ifs (to check it before `self.onset[key]` and not after) is to allow getting one element annotations. 

as in here 
https://github.com/mne-tools/mne-python/pull/5800/files#diff-0fbcacc4dac884a4e8bc277be5eaa2efR828

I think that this should hold:

```py
assert isinstance(annotations(slice(0, None, None)), Annotations)
```

**Comment by massich** [comment]
I see

File: mne/tests/test_annotations.py [hunk 7]
```diff
@@ -834,6 +834,11 @@ def test_annotations_slices():
                         description=EXPECTED_DESCS.copy(),
                         orig_time=None)
 
+    # Indexing returns a copy. So this has no effect in annot
+    annot[0]['onset'] = 42
```
Comment (massich) [inline mne/tests/test_annotations.py]: Does it surprise you that we cannot modify it? I know that this is exactly why we return a copy. But when written like this I find it surprising.

Comment (agramfort) [reply mne/tests/test_annotations.py]: I can live with that. It's fine to write:

annot.onset[0] = 42

File: mne/tests/test_annotations.py [hunk 8]
```diff
@@ -794,4 +795,69 @@ def test_read_annotation_txt_orig_time(
...
+            pass
+
+
+def test_annotations_slices():
+    """Test indexing Annotations."""
+    NUM_ANNOT = 5
+    EXPECTED_ONSETS = EXPECTED_DURATIONS = [x for x in range(NUM_ANNOT)]
+    EXPECTED_DESCS = [x.__repr__() for x in range(NUM_ANNOT)]
+
+    annot = Annotations(onset=EXPECTED_ONSETS.copy(),
```
Comment (agramfort) [inline mne/tests/test_annotations.py]: why do you need to copy?

File: tutorials/plot_object_annotations.py [hunk 9]
```diff
@@ -189,6 +180,46 @@
 print(annot)
 
 ###############################################################################
+# Iterating, Indexing and Slicing :class:`mne.Annotations`
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# :class:`~mne.Annotations` supports the same indexing and slicing as
+# ``numpy.array``, and it returns a ``(onset, duration, description)`` tuple
```
Comment (agramfort) [inline tutorials/plot_object_annotations.py]: don't forget to update this

File: mne/annotations.py (hunk 6 above)
Comment (agramfort) [reply mne/annotations.py]: you should test if key is iterable as if you pass a tuple or a numpy array of ints it should work too.

File: mne/tests/test_annotations.py (hunk 8 above)
Comment (massich) [reply mne/tests/test_annotations.py]: no reason

File: mne/tests/test_annotations.py (hunk 7 above)
Comment (massich) [reply mne/tests/test_annotations.py]: then this should be in the tuto

File: tutorials/plot_object_annotations.py (hunk 9 above)
Comment (massich) [reply tutorials/plot_object_annotations.py]: yes

File: mne/annotations.py [hunk 10]
```diff
@@ -224,30 +224,17 @@ def __iter__(self):
...
-            raise
+        if isinstance(key, int):
+            return OrderedDict(zip(
+                ('onset', 'duration', 'description', 'orig_time'),
+                (self.onset[key], self.duration[key],
+                    self.description[key], self.orig_time)))
         else:
-            return out
+            key = list(key) if isinstance(key, tuple) else key
+            return Annotations(onset=self.onset[key],
```
Comment (agramfort) [inline mne/annotations.py]: if key is a slice you should force the copy of onset etc.

Comment (massich) [reply mne/annotations.py]: Doesn't it take a copy by calling the constructor? 
I've to check that.

**Comment by agramfort** [comment]
no copy when it's a slice

>

**Comment by massich** [comment]
I guess I'm testing something wrong, because I don't see the effect of the copy.

```py
    NUM_ANNOT = 5
    EXPECTED_ONSETS = EXPECTED_DURATIONS = [x for x in range(NUM_ANNOT)]
    EXPECTED_DESCS = [x.__repr__() for x in range(NUM_ANNOT)]

    annot = Annotations(onset=EXPECTED_ONSETS,
                        duration=EXPECTED_DURATIONS,
                        description=EXPECTED_DESCS,
                        orig_time=None)

    print((id(annot.onset),
           id(annot[:1].onset),
           id(annot.onset) == id(annot[:1].onset)))
    print((id(annot.onset[0]),
           id(annot[:1].onset[0]),
           id(annot.onset[0]) == id(annot[:1].onset[0])))
    print(annot.onset[0])    NUM_ANNOT = 5
    EXPECTED_ONSETS = EXPECTED_DURATIONS = [x for x in range(NUM_ANNOT)]
    EXPECTED_DESCS = [x.__repr__() for x in range(NUM_ANNOT)]

    annot = Annotations(onset=EXPECTED_ONSETS,
                        duration=EXPECTED_DURATIONS,
                        description=EXPECTED_DESCS,
                        orig_time=None)

    print((id(annot.onset),
           id(annot[:1].onset),
           id(annot.onset) == id(annot[:1].onset)))
    print((id(annot.onset[0]),
           id(annot[:1].onset[0]),
           id(annot.onset[0]) == id(annot[:1].onset[0])))
    print(annot.onset[0])
    print((id(annot.onset[0]),
           id(annot[:1].onset[0]),
           id(annot.onset[0]) == id(annot[:1].onset[0])))
    annot[:1].onset[0] = 42
    print(annot.onset[0])
    print((id(annot.onset[0]),
           id(annot[:1].onset[0]),
           id(annot.onset[0]) == id(annot[:1].onset[0])))
    print((id(annot.onset[0]),
           id(annot[:1].onset[0]),
           id(annot.onset[0]) == id(annot[:1].onset[0])))
    annot[:1].onset[0] = 42
    print(annot.onset[0])
    print((id(annot.onset[0]),
           id(annot[:1].onset[0]),
           id(annot.onset[0]) == id(annot[:1].onset[0])))
```

the result is both the same using copy or not.
```
(139650719663168, 139650338620800, False)
(139650385747040, 139650385747040, True)
0.0
(139650385747040, 139650385747040, True)
0.0
(139650385747040, 139650385747040, True)
```
The returned list is a different one, but the elements inside are the same. But the change has no effect. I guess I'm doing something wrong. I just added the `copy()` to make sure.

**Comment by massich** [comment]
which is different than this:
```py
xx = np.array(range(3))
xx[:1][0]=42
print(xx)
```

```
[42  1  2]
```

**Comment by larsoner** [comment]
> I could not figure out how to write a test that actually breaks if .copy() is not there and passes otherwise.

That's because there is an implicit copy in the `np.array(...)` calls in the constructor:

https://github.com/mne-tools/mne-python/blob/master/mne/annotations.py#L157

So you shouldn't need to do any `.copy()` calls if the variables are passed to `Annotations(...)`

File: mne/annotations.py [hunk 11]
```diff
@@ -214,6 +214,25 @@ def __iadd__(self, other):
...
+        if isinstance(key, int):
+            out_keys = ('onset', 'duration', 'description', 'orig_time')
+            out_vals = (self.onset[key], self.duration[key],
+                        self.description[key], self.orig_time)
+            return OrderedDict(zip(out_keys, out_vals))
+        else:
+            key = list(key) if isinstance(key, tuple) else key
+            return Annotations(onset=self.onset[key].copy(),
+                               duration=self.duration[key].copy(),
+                               description=self.description[key].copy(),
```
Comment (larsoner) [inline mne/annotations.py]: ... so no need for any of these copy calls, because a copy is made in `Annotations.__init__`

**Comment by agramfort** [comment]
good catch !

**Comment by massich** [comment]
I did not see this one. 
```py
self.onset = np.array(onset, dtype=float)
```

Great.

**Comment by massich** [comment]
This still remains though
```
(139650719663168, 139650338620800, False)
(139650385747040, 139650385747040, True)
0.0
(139650385747040, 139650385747040, True)
0.0
(139650385747040, 139650385747040, True)
```
The vectors do have different id (they are indeed a copy) but the first element of both vectors share the id. Anyway.. I'll let it be. this outsmarts me.

File: doc/whats_new.rst [hunk 12]
```diff
@@ -19,6 +19,8 @@ Current
 Changelog
 ~~~~~~~~~
 
+- Add support for indexing and slicing :class:`mne.Annotations` by `Joan Massich`_
```
Comment (cbrnr) [inline doc/whats_new.rst]: Also mention iterating, i.e. "Add support for indexing, slicing, and iterating :class:`mne.Annotations` by `Joan Massich`_"

File: mne/tests/test_annotations.py [hunk 13]
```diff
@@ -794,4 +795,76 @@ def test_read_annotation_txt_orig_time(
...
+        assert isinstance(elements, OrderedDict)
+        expected_values = (ii, ii, str(ii))
+        for elem, expected_type, expected_value in zip(elements.values(),
+                                                       EXPECTED_ELEMENTS_TYPE,
+                                                       expected_values):
+            assert np.isscalar(elem)
+            assert type(elem) == expected_type
+            assert elem == expected_value
+
+    # unpacking is not allowed
```
Comment (cbrnr) [inline mne/tests/test_annotations.py]: What does this mean? Unpacking like this doesn't work because there are 4 values to unpack and you only supply 3 names.

File: doc/whats_new.rst (hunk 12 above)
Comment (massich) [reply doc/whats_new.rst]: +1

File: mne/tests/test_annotations.py (hunk 13 above)
Comment (massich) [reply mne/tests/test_annotations.py]: you cannot untuple a dict. I made it bit more explicit.

Comment (cbrnr) [reply mne/tests/test_annotations.py]: But you can unpack a dict:

```
d = dict(a=1, b=2, c=3)
x, y, z = d
```

Which is probably not what the user wants though, because this returns the keys. If you want the values, you can change that to:

```
x, y, z = d.values()
```

Comment (massich) [reply mne/tests/test_annotations.py]: yes!!, you are actually right. The test makes no sense.

Comment (massich) [reply mne/tests/test_annotations.py]: I didn't find a way to bloc the usage. I guess that [we are all consenting adults](https://python-guide-chinese.readthedocs.io/zh_CN/latest/writing/style.html#we-are-all-consenting-adults)

**Comment by cbrnr** [comment]
Thanks @massich!
